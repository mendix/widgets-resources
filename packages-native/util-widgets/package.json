{
    "name": "@native-mobile-resources/util-widgets",
    "version": "0.0.1",
    "private": true,
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "scripts": {
        "start": "tsc --watch",
        "prebuild": "rm -rf ./dist",
        "build": "tsc",
        "lint": "eslint --config ../../.eslintrc.js --ext .jsx,.js,.ts,.tsx src/",
        "prepare": "npm run build"
    },
    "dependencies": {
        "eslint": "^6.1.0",
        "typescript": "^3.5.3"
    }
}
