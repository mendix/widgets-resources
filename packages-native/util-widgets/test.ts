import { ActionValue, DynamicValue, EditableValue, ValueStatus } from "mendix";

export function dynamicValue<T>(value?: T): DynamicValue<T> {
    return value == null ? { status: ValueStatus.Loading, value: undefined } : { status: ValueStatus.Available, value };
}

export function actionValue(canExecute = true, isExecuting = false): ActionValue {
    return { canExecute, isExecuting, execute: jest.fn() };
}

export class EditableValueBuilder<T extends string | boolean | Date | BigJs.Big> {
    private readonly editableValue: EditableValue<T> = {
        value: undefined,
        displayValue: "",
        status: ValueStatus.Available,
        validation: undefined,
        readOnly: false,
        formatter: {
            format: jest.fn(name => `Formatted ${name}`),
            parse: jest.fn(),
            getFormatPlaceholder: jest.fn()
        },
        setValidator: jest.fn(),
        setValue: jest.fn((value: T) => this.withValue(value)),
        setTextValue: jest.fn(),
        setFormatter: jest.fn()
    };

    withValue(value?: T): EditableValueBuilder<T> {
        this.editableValue.value = value;
        this.editableValue.displayValue = this.editableValue.formatter.format(value);
        return this;
    }

    isReadOnly(): EditableValueBuilder<T> {
        this.editableValue.readOnly = true;
        return this;
    }

    isLoading(): EditableValueBuilder<T> {
        this.editableValue.status = ValueStatus.Loading;
        return this.isReadOnly();
    }

    isUnavailable(): EditableValueBuilder<T> {
        this.editableValue.status = ValueStatus.Unavailable;
        return this.isReadOnly();
    }

    withValidation(validation?: string): EditableValueBuilder<T> {
        this.editableValue.validation = validation;
        return this;
    }

    withUniverse(...values: T[]): EditableValueBuilder<T> {
        this.editableValue.universe = values;
        return this;
    }

    build(): EditableValue<T> {
        return this.editableValue;
    }
}
