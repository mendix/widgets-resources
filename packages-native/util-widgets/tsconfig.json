{
    "extends": "@mendix/pluggable-widgets-tools/configs/tsconfig.base",
    "baseUrl": "./",
    "include": ["./src"],
    "compilerOptions": {
        "target": "es2015",
        "lib": ["es2015"],
        "types": ["mendix-client", "big.js", "react-native", "jest"],
        "moduleResolution": "node",
        "module": "es2015",
        "allowSyntheticDefaultImports": true,
        "esModuleInterop": true,
        "strict": true,
        "noEmitOnError": true,
        "noFallthroughCasesInSwitch": true,
        "noImplicitAny": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "forceConsistentCasingInFileNames": true,
        "skipLibCheck": true,
        "jsx": "react",
        "jsxFactory": "createElement",
        "baseUrl": "./",
        "outDir": "./dist",
        "declaration": true,
        "sourceMap": false
    }
}
