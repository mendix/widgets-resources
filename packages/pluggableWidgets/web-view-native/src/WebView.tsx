import { flattenStyles } from "@mendix/piw-native-utils-internal";
import { Component, createElement, createRef, RefObject } from "react";
import { Text, View, Linking, Platform, BackHandler } from "react-native";
import { WebView as RNWebView } from "react-native-webview";

import { WebViewProps } from "../typings/WebViewProps";
import { defaultWebViewStyle, WebViewStyle } from "./ui/Styles";
import { executeAction } from "@mendix/piw-utils-internal";

export type Props = WebViewProps<WebViewStyle>;

export class WebView extends Component<Props> {
    private readonly onLoadHandler = this.onLoad.bind(this);
    private readonly onErrorHandler = this.onError.bind(this);
    private readonly onMessageHandler = this.onMessage.bind(this);
    private readonly onAndroidBackPressHandler = this.onAndroidBackPress.bind(this);
    private readonly styles = flattenStyles(defaultWebViewStyle, this.props.style);

    private webViewRef: RefObject<RNWebView> = createRef();
    private canGoBack = false;

    componentDidMount(): void {
        if (Platform.OS === "android" && this.props.propagateBackbutton) {
            BackHandler.addEventListener("hardwareBackPress", this.onAndroidBackPressHandler);
        }
    }

    componentWillUnmount(): void {
        if (Platform.OS === "android" && this.props.propagateBackbutton) {
            BackHandler.removeEventListener("hardwareBackPress", this.onAndroidBackPressHandler);
        }
    }

    render(): JSX.Element {
        const uri = this.props.url && this.props.url.value;
        const html = this.props.content && this.props.content.value;

        if (!uri && !html) {
            return (
                <View style={this.styles.errorContainer}>
                    <Text style={this.styles.errorText}>No URL or content was provided.</Text>
                </View>
            );
        }

        return (
            <View style={this.styles.container}>
                <RNWebView
                    ref={this.webViewRef}
                    testID={this.props.name}
                    source={html ? { html } : { uri: uri! }}
                    style={{
                        width: "100%",
                        height: "100%"
                    }}
                    onLoad={this.onLoadHandler}
                    onError={this.onErrorHandler}
                    onMessage={event => {
                        this.onMessageHandler(event.nativeEvent.data);
                    }}
                    userAgent={this.props.userAgent}
                    onShouldStartLoadWithRequest={({ url }) => {
                        const openExternally =
                            this.props.openLinksExternally && (html ? url.slice(0, 4) === "http" : url !== uri);
                        if (openExternally) {
                            Linking.openURL(url);
                            return false;
                        }
                        return true;
                    }}
                    allowsBackForwardNavigationGestures={this.props.allowsBackForwardNavigationGestures}
                    onNavigationStateChange={navState => {
                        this.canGoBack = navState.canGoBack;
                    }}
                />
            </View>
        );
    }

    private onLoad(): void {
        executeAction(this.props.onLoad);
    }

    private onError(): void {
        executeAction(this.props.onError);
    }

    private onMessage(input: string): void {
        this.props.onMessageInput?.setTextValue(input);
        executeAction(this.props.onMessage);
    }

    private onAndroidBackPress(): boolean {
        if (this.canGoBack && this.webViewRef.current) {
            this.webViewRef.current.goBack();
            return true;
        } else {
            return false;
        }
    }
}
