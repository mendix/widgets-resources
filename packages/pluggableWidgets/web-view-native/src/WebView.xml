<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.webview.WebView" supportedPlatform="Native" needsEntityContext="true" offlineCapable="true" pluginWidget="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Web view</name>
    <description>Display an external web page or load custom HTML.</description>
    <icon>iVBORw0KGgoAAAANSUhEUgAAAEAAAABACAYAAACqaXHeAAAOsklEQVR4Xu1bfZQcVZX/3Vszk2QmJtH1HD5EpnsCBAiiLuyCyCII4h6WGAx0dYsYUAIciAgh3Z0EQhgR8tHdE4IYVtgAAZVMdycqRl0Qkc/IIq4cXEIymHR1B1Si8pVkEiY99e6equ7qqe5Ud9cMQ/QcqL+mq+59973fu/e++/GG8B5/6D2+frwPwPsa8B5HYL+ZwGFXfn9CcUzbVCJMYZKAEp5AhPEl/M2dorQdYJUHpG8Mxm3sS0zfuT/25l0FIBDvPZEUZoD5szDxSTDY36KUqYSf00CPCMsPjYT+G398w6cadQCmxB/4wIDsvVxILmbBEcOfkheH2gzSVrW3yx0bu/VdozNmaZRRA+CA6H0dHTwmJiJXATxpNCc5NJZ6HUIriialXrlF3zMaMkYFgEA8fY4Myq2s8aG1kxKl/kCEH4tGj8DEFCK6tdHECbhCAQZBzhDQOQRM3odeIQ+Sbxip8Pp3CsI7AuDg7vXtY/v7Vwr4IvdEFFBkUb3CuD2fiPyP9S0Yz0zBoPodNG5vNGklalcr4RNbkpGtFt3kedmTTGXOFgWdmVuq5cgq6th9Zb77q2+PFIgRA9AZzQRJ1HpinuoWLlD3i/DCQko3Ku9DGS0QwNME/Iv1zgIIwCYGjrV+C9T/icJRrgVuMDpePAXd3coZY/L8zGGqqG4Cc7gG7N9rjGm5Zfq2kYAwIgCC0bXHgs0HIXSQyz5zItrF+VTosdqJBOPZuRBJVd4T4lA0FSQX2u+E7iVGn4gsdtFcaST079SO1RXPni5i3gVwp/ONoP7I3PL5LcvO2zhcEIYNgL14Gny8ytERZdrQNsvr7O68et1B0Ip9zPwBe/fFfKowvu8zwV1T73YDYIzf+LVg/9EbAJxo0ym81dqGw7cs0f+6DwjzMhOhsFqAc9wOUuOWU4YLwrAAsNSeWTa4d56gbswlIzfUQz4Y710F4Ysrqk/q44VEZFMwml1dBUAqdNHkaO8xg4LnKqYgcruRCs/2HlsoEEsvJvB8tyaA+aThmINvACyH17az/zdumyfB3FxKX15v8bbdmmozwFqJRnqMZDhq/eUFgP0+nr0VIt9wAGOlHW70nFuoJ6MrlpkvwBLnuwJ+Tx39J/h1jL4BCETTdxPRV4dslL5lJEKLGtlcMJ69EyKXOCotrcXgtqVffqMRAEfMvf/DA+CcYzIArTSSoa83khOI9S5xawKg7jSSkcv8+ANfAASi2elE8uPKgAp5IVrQSAAzNDFlFRhjy7v5SxK6q6KuLJdDcIr9m/CEKPpP5xuLXCqM0+zfptpNzJcoUOVEqJXLpMg0Zak7DiHB2bmU/rNmIDQFwFL91h39m7yCnGaD/32/K6M4yFObRYxNAQhEszcQSfffdzEjlC5YaKT0mxtxNwTASmz2mgMFMD5YVuMdAJrG4Aw1xn1MKlF/BXGVCrPCRMc8oPC2YrxVM1GNgQ8PvVNvKvCADyjGMTDBplPma7up2Lk9NbO/Hl9DAILxTAyCRMmJqUEQH1EV4dUZNRjNXAfCTWXQthfy+Aiyuukmr3cKODSndj/aUujfvh3gD5X9RNxI6MlmAHTF1h4uGNzknDzNTqrGAMzt3QzmKSWhcp+RDJcityZPZ3TNk0zaySUAZFUhGbZPguEAYNEGo733gfgrZb5HjKR+RjPZNt/c3l4nZCaoF3PJSFW47h6jLgCB6NoTiJSdyJR34AQ/hYlD5mTGaazeZOa2MqNuJEPZkQGQvgBE3yvxqj3tHTxpY7e+txkIXfMyJ4vCkw4dKxy/tUf/Xy++ugB0xdLLBBQvyVZ9Rk/kyGaCre+TY+lPK9BTFdrWwQONxedvHwkAXfMyh4pCJQgipf411xN51s88grHeHMBBi1ZEluRT4WuHBUAwnvktBMdZTCS0LJcKVULORhMIxtKzAbKTGGWqbYXlkUrSMlwTsOg7Y5lXGTigrIWXGQn9Tj8ABGLZ5QSZU6Z9xkjqdo5R+3hqgFXANNva3nBqeER0Ri4ResSf4N7bCXx5SXHw00KPPs2LryvWe0+ljiB0r5EKVdUUHJ5ALP0LAn3O3giSFblE2FlUw+l0xdJnCagcCCmzbD77lNM8AbCLmcJPlyQoaaNxE/1WaYOxzEMAziztmFpuJCJzvWYamLv2SNbkOBEcDVHPG6lwxosuGMt+BxA7IRJgfT6pf8HPRhwyJ/Oh1ha85tDWMx9vAGKZiwi4p7R+5I0e3bYlP09gbu8LTsIkwNX5pN6wBNZsTPdRTCZ+l1uu22bp5wnG03+qZK6EmUZCLzvUIW5PALri6W4RclLcx4ykXorLfTxumxWhL+VToV4vtqndmbY9e3Bgs9Q1GMvOBOReawwTeHlbUt+n7lhvWoFoZgMRTrK/iywyUuFv1dLW0YAhB2I5MmbtUR9rLykMyfkMtJaE4nGQvFKegNIGJlyy5bazBgLz1wSkyE8y0yECPEughUYy9ItaGZ3x9GkM6naSJgU1wKJ5Auo1PyXqdEtG6RuljGQo5gsAdxrrd+HN6ZRpJCN2UTMQyywi4JtVPIQniHBdbpleOUIDV/9oErfuXSCmfL1ZMdWHfM8U2VMD3h0AAO3tvRO33HbBjkAsEyagzk7Kg6xooTtwCcQyBxLkWqXksqEAq/mSqwFWdxmJyCyfGtDbA+FrbJV+ByYgkIeJxK7WKiFpMdWNW5d/6WWEMlrwULVxKMzedzEE+RGzdr27xjf5mjUfFU37shB1AWbTNpspfKYGfLQ0utxmJMN2pcn9ePuA6hR4xE6QIOfnkuE1XnsVjLrD3Dq7qaDAskaD3OD0CYaz724nKELd+VSo2uzqtcYC8fSFJLS6JEwVjGQk4Fdw1TEoNCefCq3w5PWhBQ6fnYkyra5okM/JBGO9fwb4QNsFQi7OJcN3+9OAmkBIe3twkmW7fuQGo5kHQfi8rXSgW/LJkG1KfrTAOuY0yL1KyVVDNcEhTvsUIO0OtBQXe+UXbhlWbbHILZWSOjN9euuy0K99AWCHwmNbXq9UcwlnGgn9YX8ApFeC6AqbluRnRiJ8dl0+Ly0QCbeK+asit84DzNkAj/Pg7xeo25RmJpwiay1NVyx7tkDKvUNl7pa9E70KI3Wzwc5o5lkmHG8PLCpppCKlzLDJ0xXPXiEiK23jUfJKoSdcdkL+tECU2pjvCX/MQs9qqnDL4EIlMsvL+1vNE2IsH0NjbqkN1btivSsEfJW9Dw0iyPrpcDS7VEjmldav/pDvifjq9VvNTKXE6vDYjypqBxdWnPvn4WqBOzewAicM8iIimTnUY3CbBv5GwmfnU+c9U3orFIylDad91iib9V0QEVKfcjq9jZQg0H3PWNk57i1nxwSI5JN6uhFP7Yng1gI3n91hNtU3wdABrp37OiOpn2fRB6LZU4lkKHol1C3mDKck9gMjGb6gmQlY34PxzONO+CqK7s73hOzWWNVioulpJPKqXeCo4wu8M0Rrd7NWhelc93juVDkYy6ytfG9SzGkMQCwdBcguRFpHUSvjSD/ncVc8u8Dp9FoV4UKBD9qnKBpLfx+gyUZS/5QNWk1c4KkF3d0c3D31u063yQXAY+0dmGZdnym10dHn1DKaZaTNy+KyJ1+pzAL9UGbTywgK1FZ1jCm8YbV43DummMYzeAwLTt+a0n/VVAtCGa0zQKsZUq2FgoeKJr5oN0AsTQrITwH697IHer29gzsb3Stq2hgJRtPXg+hGP6o/EhpSeDTXo3+2kRYcd+mdLX+bMOl+JrJtfOiRn7R3UMgplLo70RYNAQtySX1po3k1BcCq8rYyXgTDdzQ4XCCU4KRCSn/aUwsIM5VpXY/BPvGElWsMDtJ0p/1VVWaDMqRjz9HNusRNASjvzDQQ/aSyMOuSElMcJFJvsSLQCOoeJ5ApJUY0VNAUWGWuU+1DyxUx7pMj2PlA/fuFbhCsZoqx6y9rLE1Ris4q9IT+u9lm+AKg5NmHLjqUJq2W5pORhh3irlj6uwKy29RKqZ2DbW3BPy6ZYdfpqjpD5EpVm+QIVlAjGv65yhDcmmDxd6ljjGWR55stvmwmfsgAyxRaWtQzBP6Yw0FE1+YSocrlhNqRrBslEPWSc+ODoG7NJSNXNwTA40SojFsusu57H8DakGpz8LeqYV6UDM5d10lc3CDgj1RAsHsG5y2wQlcvocFY7x0AX2prAVBsZf6klePX1QCL0EsLSN1sJCILHRmBaHox1dxRGAkIvk3AEWzd41HCTzgd4/L7B4hxYW6ZXtvhRfDa+w9QxZaXKh1bmE8bHX0nV12SskzA4MsCQbkg377pe9b1OOs2mGnKOkB1aMzXe3nz0QBh2ABYCy6D8BAYBzvA2JWjFp7llTUGYpmrCKjUBSzTEYUplUtSpO4S4W1WnZCIvpZLhOyS/GFX/nyMNq7Y5k50grHsv7FofVtTM/5im1IsezMgVW0vSxPQsfsLzU6AYfmAWvW2zEG4uN7tE2w1F1mrKbkutzzyUoXHiuD6pzwFaHbUVypw8GYGjimbxgsMdVQ50elnUSduTUVecMu0GylsLhbQFwG1WcCn5ZP6q2UHfROEr3PT++0ijUgDHEF2J7hFvs2gmmKjMkHaOphqpdGjP2n5BytEHRxUzzFx+X8EvN2UAnaw0o4t3QwTCkTXfkagZrPQDPdxqAS/LaR0++apFwgi+Hk+pf9HM2f4jgBwBi8VH8xvO93YaqGqANIegFK/FCDo57K0iLwCwueUKdO97iZZ5qYxX1F7Ccq+ziOySJHa3QKebofYTZ5RAcCSYUeMGq6BmHPA2j81Ezyy7+pNQOsZ6Bi7/E/d03Z7jXHo/B98cHDshIF632t5Rg0AZ+Cp3Znxe3bhUgFmgXDUyBZazWVfude0O7Q9A//ltzbpV+6oA+AW3BVfe7xS5gyrRaUxHedVzfGcqBX+anieRB5Wov1wqNLjd1n+6d5VANzTsG+cYcDSiClQEgDxBMCcIEprIVb9Ivwak3qZgL6xHfT8aP9rTD1I9hsA/vdk/1K+D8D+xfsfT9r7GvCPtyf7d0b/D9ViGqrC0c7OAAAAAElFTkSuQmCC</icon>
    <properties>
        <propertyGroup caption="Data source">
            <property key="url" type="textTemplate" required="false">
                <caption>URL</caption>
                <description>The URL of the web page that is loaded.</description>
            </property>
            <property key="content" type="textTemplate" required="false">
                <caption>Content</caption>
                <description>The HTML content to be rendered. This will precede URL</description>
            </property>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onLoad" type="action" required="false">
                <caption>On load</caption>
                <description/>
            </property>
            <property key="onError" type="action" required="false">
                <caption>On error</caption>
                <description/>
            </property>
        </propertyGroup>
        <propertyGroup caption="Advanced">
            <property key="userAgent" type="string" required="false">
                <caption>User agent</caption>
                <description>A custom user agent string will be included with the request.</description>
            </property>
            <property key="openLinksExternally" type="boolean" defaultValue="false">
                <caption>Open links externally</caption>
                <description>If true, links embedded in source document will be opened in external browser instead of inside this widget.</description>
            </property>
        </propertyGroup>
    </properties>
</widget>
