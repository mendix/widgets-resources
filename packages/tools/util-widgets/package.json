{
  "name": "@native-mobile-resources/util-widgets",
  "version": "1.0.0",
  "private": true,
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "start": "..\"/../../node_modules/.bin/tsc\" --watch",
    "prebuild": "..\"/../../node_modules/.bin/rimraf\" ./dist",
    "build": "..\"/../../node_modules/.bin/tsc\"",
    "lint": "..\"/../../node_modules/.bin/eslint\" --config ../../../.eslintrc.js --ext .jsx,.js,.ts,.tsx src/",
    "prepare": "npm run build"
  },
  "devDependencies": {
    "@mendix/pluggable-widgets-tools": ">=8.9.2"
  }
}
