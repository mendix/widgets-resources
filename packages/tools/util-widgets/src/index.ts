export { flattenStyles, Style } from "./styles";
export { only, exclude } from "./objectFilters";
export { available, toNumber, unavailable } from "./properties";
export * from "./test";
