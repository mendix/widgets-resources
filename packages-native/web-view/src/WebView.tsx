import { flattenStyles } from "@native-mobile-resources/util-widgets";
import { Component, createElement } from "react";
import { Text, View } from "react-native";
import { WebView as RNWebView } from "react-native-webview";

import { WebViewProps } from "../typings/WebViewProps";
import { defaultWebViewStyle, WebViewStyle } from "./ui/Styles";
import { executeAction } from "@widgets-resources/piw-utils";

export type Props = WebViewProps<WebViewStyle>;

export class WebView extends Component<Props> {
    private readonly onLoadHandler = this.onLoad.bind(this);
    private readonly onErrorHandler = this.onError.bind(this);
    private readonly styles = flattenStyles(defaultWebViewStyle, this.props.style);

    render(): JSX.Element {
        const uri = this.props.url && this.props.url.value;
        const html = this.props.content && this.props.content.value;

        if (!uri && !html) {
            return (
                <View style={this.styles.errorContainer}>
                    <Text style={this.styles.errorText}>No URL or content was provided.</Text>
                </View>
            );
        }

        return (
            <View style={this.styles.container}>
                <RNWebView
                    testID={this.props.name}
                    source={html ? { html } : { uri: uri! }}
                    style={{
                        width: "100%",
                        height: "100%"
                    }}
                    onLoad={this.onLoadHandler}
                    onError={this.onErrorHandler}
                    userAgent={this.props.userAgent}
                    useWebKit
                />
            </View>
        );
    }

    private onLoad(): void {
        executeAction(this.props.onLoad);
    }

    private onError(): void {
        executeAction(this.props.onError);
    }
}
