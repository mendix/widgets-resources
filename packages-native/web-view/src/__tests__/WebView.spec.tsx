import { actionValue, dynamicValue } from "@native-mobile-resources/util-widgets";
import { createElement } from "react";
import { fireEvent, render } from "react-native-testing-library";
import { WebView as RNWebView } from "react-native-webview";

import { Props, WebView } from "../WebView";

describe("WebView", () => {
    let defaultProps: Props;

    beforeEach(() => {
        defaultProps = {
            name: "webview-test",
            style: [],
            url: dynamicValue(false, "https://mendix.com")
        };
    });

    it("renders a web view when a url is provided", () => {
        const component = render(<WebView {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders a web view when html content is provided", () => {
        const component = render(<WebView {...defaultProps} content={dynamicValue(false, "Hello, world!")} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders an error when no url or content is provided", () => {
        const component = render(
            <WebView {...defaultProps} url={dynamicValue<string>(true)} content={dynamicValue<string>(true)} />
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("executes the on load action", () => {
        const onLoadAction = actionValue();
        const component = render(<WebView {...defaultProps} onLoad={onLoadAction} />);

        fireEvent(component.getByType(RNWebView), "load");

        expect(onLoadAction.execute).toHaveBeenCalledTimes(1);
    });

    it("executes the on error action", () => {
        const onErrorAction = actionValue();
        const component = render(<WebView {...defaultProps} onError={onErrorAction} />);

        fireEvent(component.getByType(RNWebView), "error");

        expect(onErrorAction.execute).toHaveBeenCalledTimes(1);
    });
});
