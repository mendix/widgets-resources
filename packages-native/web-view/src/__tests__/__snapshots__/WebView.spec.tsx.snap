// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebView renders a web view when a url is provided 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": "100%",
      "minHeight": 300,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RNCWKWebView
      cacheEnabled={true}
      messagingEnabled={false}
      onError={[Function]}
      onLoad={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      source={
        Object {
          "uri": "https://mendix.com",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      testID="webview-test"
      useSharedProcessPool={true}
    />
  </View>
</View>
`;

exports[`WebView renders a web view when html content is provided 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": "100%",
      "minHeight": 300,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "overflow": "hidden",
      }
    }
  >
    <RNCWKWebView
      cacheEnabled={true}
      messagingEnabled={false}
      onError={[Function]}
      onLoad={[Function]}
      onLoadingError={[Function]}
      onLoadingFinish={[Function]}
      onLoadingProgress={[Function]}
      onLoadingStart={[Function]}
      onMessage={[Function]}
      onShouldStartLoadWithRequest={[Function]}
      source={
        Object {
          "html": "Hello, world!",
        }
      }
      style={
        Array [
          Object {
            "flex": 1,
            "overflow": "hidden",
          },
          Object {
            "backgroundColor": "#ffffff",
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
        ]
      }
      testID="webview-test"
      useSharedProcessPool={true}
    />
  </View>
</View>
`;

exports[`WebView renders an error when no url or content is provided 1`] = `
<View
  style={Object {}}
>
  <Text
    style={
      Object {
        "color": "red",
        "fontWeight": "bold",
      }
    }
  >
    No URL or content was provided.
  </Text>
</View>
`;
